File: solutions/code-carl/ch06_tree/t22_lc0530_get_minimum_difference.py
"""LeetCode 530. Minimum Absolute Difference in BST
https://leetcode.cn/problems/minimum-absolute-difference-in-bst/
"""
import sys
from typing import Optional

from tree_node import TreeNode


class Solution:
    pre_num = -sys.maxsize - 1
    min_diff = sys.maxsize

    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        def traversal(root: Optional[TreeNode]) -> None:
            if not root:
                return
            traversal(root.left)
            if self.pre_num != -sys.maxsize - 1:
                self.min_diff = min(self.min_diff, root.val - self.pre_num)
            self.pre_num = root.val
            traversal(root.right)

        traversal(root)
        return self.min_diff if self.min_diff != sys.maxsize else -1


if __name__ == "__main__":
    solution = Solution()
    root1 = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(6))
    assert solution.getMinimumDifference(root1) == 1

    solution = Solution()
    root2 = TreeNode(1, TreeNode(0), TreeNode(48, TreeNode(12), TreeNode(49)))
    assert solution.getMinimumDifference(root2) == 1

    print("All tests passed.")

File: solutions/java/code-carl/ch06_tree/T22_LC0530_GetMinimumDifference.java
/**
 * Practice template for:
 * - LeetCode 530. Minimum Absolute Difference in BST
 *   https://leetcode.cn/problems/minimum-absolute-difference-in-bst/
 */
public class T22_LC0530_GetMinimumDifference {
    Integer preVal;
    Integer minDiff = 0x3f3f3f3f;

    public int getMinimumDifference(TreeNode root) {
        if (root == null) {
            return -1;
        }
        traversal(root);
        return minDiff;
    }

    public void traversal(TreeNode root) {
        if (root == null) {
            return;
        }
        traversal(root.left);
        if (preVal != null) {
            minDiff = Math.min(minDiff, root.val - preVal);
        }
        preVal = root.val;
        traversal(root.right);
    }
}
